db/migrations/20210114161425_update_urn_snapshot_trigger.sql

```diff
@@ -0,0 +1,506 @@
+        PERFORM maker.update_urn_inks_until_next_diff(NEW, NEW.ink);
+        PERFORM maker.update_urn_created(NEW.urn_id);
+    ELSIF (TG_OP = 'DELETE') THEN
+        PERFORM maker.update_urn_inks_until_next_diff(OLD, urn_ink_before_block(OLD.urn_id, OLD.diff_id));
+        PERFORM maker.delete_obsolete_urn_snapshot(OLD.urn_id, OLD.header_id, OLD.diff_id);
+        PERFORM maker.update_urn_created(OLD.urn_id);
+    END IF;
+    RETURN NULL;
+END
+$$
+    LANGUAGE plpgsql;
+-- +goose StatementEnd
+
+-- update insert_urn_art function to use new urn_ink_before_block function
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.insert_urn_art(new_diff maker.vat_urn_art) RETURNS maker.vat_urn_art
+AS
+$$
+BEGIN
+    WITH urn AS (
+        SELECT urns.identifier AS urn_identifier, ilks.identifier AS ilk_identifier
+        FROM maker.urns
+                 LEFT JOIN maker.ilks ON urns.ilk_id = ilks.id
+        WHERE urns.id = new_diff.urn_id),
+         new_diff_header AS (
+             SELECT api.epoch_to_datetime(block_timestamp) AS block_timestamp, block_number
+             FROM public.headers
+             WHERE id = new_diff.header_id)
+    INSERT
+    INTO api.urn_snapshot (urn_identifier, ilk_identifier, block_height, ink, art, created, updated)
+    VALUES ((SELECT urn_identifier FROM urn),
+            (SELECT ilk_identifier FROM urn),
+            (SELECT block_number FROM new_diff_header),
+            urn_ink_before_block(new_diff.urn_id, new_diff.diff_id),
+            new_diff.art,
+            urn_time_created(new_diff.urn_id),
+            (SELECT block_timestamp FROM new_diff_header))
+    ON CONFLICT (urn_identifier, ilk_identifier, block_height)
+        DO UPDATE SET art = new_diff.art;
+    RETURN NULL;
+END
+$$
+    LANGUAGE plpgsql;
+-- +goose StatementEnd
+
+-- update update_urn_arts to use the new delete_obsolete_urn_snapshot function
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.update_urn_arts() RETURNS TRIGGER
+AS
+$$
+BEGIN
+    IF (TG_OP IN ('INSERT', 'UPDATE')) THEN
+        PERFORM maker.insert_urn_art(NEW);
+        PERFORM maker.update_urn_arts_until_next_diff(NEW, NEW.art);
+    ELSIF (TG_OP = 'DELETE') THEN
+        PERFORM maker.update_urn_arts_until_next_diff(OLD, urn_art_before_block(OLD.urn_id, OLD.diff_id));
+        PERFORM maker.delete_obsolete_urn_snapshot(OLD.urn_id, OLD.header_id, OLD.diff_id);
+    END IF;
+    RETURN NULL;
+END
+$$
+    LANGUAGE plpgsql;
+-- +goose StatementEnd
+
+-- update update_urn_arts_until_next_diff to use the snapshot's diff instead of its header to get block_number
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.update_urn_arts_until_next_diff(start_at_diff maker.vat_urn_art, new_art NUMERIC) RETURNS maker.vat_urn_art
+AS
+$$
+DECLARE
+    diff_block_number    BIGINT := (
+        SELECT block_height
+        FROM public.storage_diff
+        WHERE id = start_at_diff.diff_id);
+    next_rate_diff_block BIGINT := (
```

Comment:
nbd but this could probably be `next_art_diff_block`, I think?

Reply:
Nice catch!